lsp/src/text_document.ml

```diff
@@ -1,114 +1,89 @@
 open Types
+module String = StringLabels
 
-module Encoding = struct
-  let recode ?nln ?encoding out_encoding
-      (src : [ `Channel of in_channel | `String of string ])
-      (dst : [ `Channel of out_channel | `Buffer of Buffer.t ]) =
-    let rec loop d e =
-      match Uutf.decode d with
-      | `Uchar _ as u ->
-        ignore (Uutf.encode e u);
-        loop d e
-      | `End -> ignore (Uutf.encode e `End)
-      | `Malformed _ ->
-        ignore (Uutf.encode e (`Uchar Uutf.u_rep));
-        loop d e
-      | `Await -> assert false
-    in
-    let d = Uutf.decoder ?nln ?encoding src in
-    let e = Uutf.encoder out_encoding dst in
-    loop d e
-
-  let nln = `ASCII (Uchar.of_char '\n')
-
-  let reencode_string in_enc out_enc str =
-    let buf = Buffer.create (String.length str) in
-    let () = recode ~nln ~encoding:in_enc out_enc (`String str) (`Buffer buf) in
-    Buffer.contents buf
-
-  let utf8_to_utf16 = reencode_string `UTF_8 `UTF_16LE
-
-  let utf16_to_utf8 = reencode_string `UTF_16LE `UTF_8
+exception Invalid_utf8
 
-  let utf16_line_offsets (text : string) =
-    let rec loop d acc =
-      let old_line = Uutf.decoder_line d in
-      match Uutf.decode d with
-      | `Uchar _
-      | `Malformed _ ->
-        let new_line = Uutf.decoder_line d in
-        if new_line > old_line then
-          let line_ofs = Uutf.decoder_byte_count d / 2 in
-          (* UTF16 encodes on 2 bytes *)
-          loop d (line_ofs :: acc)
-        else
-          loop d acc
-      | `End ->
-        let end_ofs = Uutf.decoder_byte_count d / 2 in
-        Array.of_list (List.rev (end_ofs :: acc))
-      | `Await -> assert false
-    in
-    let encoding = `UTF_16LE in
-    let decoder = Uutf.decoder ~nln ~encoding (`String text) in
-    loop decoder [ 0 ]
-end
+let find_offset ~utf8 ~utf16_range:range =
+  let dec =
+    Uutf.decoder
+      ~nln:(`ASCII (Uchar.of_char '\n'))
+      ~encoding:`UTF_8 (`String utf8)
+  in
+  let bytes = Bytes.create 4 in
+  let enc = Uutf.encoder `UTF_16LE `Manual in
+  let rec find_char line char =
+    if char = 0 || Uutf.decoder_line dec = line then
+      Uutf.decoder_byte_count dec
+    else
+      match Uutf.decode dec with
+      | `Await -> raise Invalid_utf8
+      | `End -> Uutf.decoder_byte_count dec
+      | `Malformed _ -> assert false
```

Comment:
We assert false because we assume that the current state of the text document is a valid utf8 string. If it was invalid, we should have raised when it became invalid and not here.